mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java

```diff
@@ -0,0 +1,494 @@
+    private CloseableIterable<T> newAvroIterable(InputFile inputFile, FileScanTask task, Schema readSchema) {
+      Avro.ReadBuilder avroReadBuilder = Avro.read(inputFile)
+                                             .project(readSchema)
+                                             .split(task.start(), task.length());
+
+      if (reuseContainers) {
+        avroReadBuilder.reuseContainers();
+      }
+
+      switch (inMemoryDataModel) {
+        case PIG:
+        case HIVE:
+          //TODO implement value readers for Pig and Hive
+          throw new UnsupportedOperationException();
+        case DEFAULT:
+          avroReadBuilder.createReaderFunc(DataReader::create);
+      }
+      return avroReadBuilder.build();
+    }
+
+    private CloseableIterable<T> newParquetIterable(InputFile inputFile, FileScanTask task, Schema readSchema) {
+      Parquet.ReadBuilder parquetReadBuilder = Parquet.read(inputFile)
+                                                      .project(readSchema)
+                                                      .filter(task.residual())
+                                                      .caseSensitive(caseSensitive)
+                                                      .split(task.start(), task.length());
+      if (reuseContainers) {
+        parquetReadBuilder.reuseContainers();
+      }
+
+      switch (inMemoryDataModel) {
+        case PIG:
+        case HIVE:
+          //TODO implement value readers for Pig and Hive
+          throw new UnsupportedOperationException();
+        case DEFAULT:
+          parquetReadBuilder.createReaderFunc(
+              fileSchema -> GenericParquetReaders.buildReader(readSchema, fileSchema));
+      }
+      return parquetReadBuilder.build();
+    }
+
+    private CloseableIterable<T> newOrcIterable(InputFile inputFile, FileScanTask task, Schema readSchema) {
+      ORC.ReadBuilder orcReadBuilder = ORC.read(inputFile)
+                                          .schema(readSchema)
+                                          .caseSensitive(caseSensitive)
+                                          .split(task.start(), task.length());
+      // ORC does not support reuse containers yet
+      switch (inMemoryDataModel) {
+        case PIG:
+        case HIVE:
+          //TODO implement value readers for Pig and Hive
+          throw new UnsupportedOperationException();
+        case DEFAULT:
+          //TODO: We do not have support for Iceberg generics for ORC
+          throw new UnsupportedOperationException();
+      }
+
+      return orcReadBuilder.build();
+    }
+  }
+
+  private static Table getTable(Configuration conf) {
+    String path = conf.get(TABLE_PATH);
+    if (path.contains("/")) {
+      HadoopTables tables = new HadoopTables(conf);
+      return tables.load(path);
+    } else {
+      Catalog catalog = HiveCatalogs.loadCatalog(conf);
+      TableIdentifier tableIdentifier = TableIdentifier.parse(path);
+      return catalog.loadTable(tableIdentifier);
```

Comment:
For a simple MR application, do we need to involve `HiveCatalog`?

It seems strange to me that a simple MR application requires to communicate with Hive. I was thinking that if we could have an abstract catalog layer here, or only support Hadoop tables in MR as default implementation. Besides let Hive, Pig and others to choose a specific implementation, because each engine may have their own preference, and we don't have to bind to Hive or others here.

Reply:
I like this idea, but I don't think it works. The problem is that `getSplits` isn't called in the client/driver process. It's called in the setup task that needs to instantiate the `InputFormat` with the no-arg constructor. So everything needs to be passed through the `Configuration`.

Tables themselves aren't serializable, so we can't pass a table or a scan that way. We need to be able to load the table in the setup task, configure the scan, and get splits there.